Include Column Lineage Info for All Queries

behrooz-stripe opened the pull request:
## Description
Column lineage info is extremely helpful but is only available for Create View, Table etc. not Select queries. This PR simply expands the support for lineage for other queries as well.


## Additional context and related issues
adds column lineage to updateTarget so it'll be available in `EventListener` without any other plumbing/changes required. Please see `assertLineage` for testing, it'll force all the select queries to go through the same tests and assumptions.


## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
(x) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

behrooz-stripe commented:
Hi, I work for [Stripe](https://stripe.com/en-ca) and I've double checked on our end and we already should have a Corporate CLA signed. Please verify on your end. Thanks!

mosabua commented:
@behrooz-stripe you still need to sign the individual contributor license agreement

behrooz-stripe commented:
> @behrooz-stripe you still need to sign the individual contributor license agreement

thanks @mosabua ! I just emailed the individual one too.

wendigo commented:
@behrooz-stripe Can you squash commits so that the commit is a logical unit of a change that fully compiles on its own?

behrooz-stripe commented:
@wendigo done.

wendigo commented:
@behrooz-stripe Please drop merge commit and rebase on master and I'll review it

behrooz-stripe commented:
> @behrooz-stripe Please drop merge commit and rebase on master and I'll review it

alright, should be good now! I think I accidentally clicked on sync button in GitHub UI an it committed the merge one.

behrooz-stripe commented:
@wendigo seems like after I rebased this `testCacheFileOperations` is failing but I can't seem to be able to reproduce it locally and it succeeds. Could it be it's flakey? 
<img width="694" alt="image" src="https://github.com/user-attachments/assets/8790120a-de5f-4695-a58e-103c54e0a0fa">

behrooz-stripe commented:
nvm, I can see the master is failing with the same error too.

behrooz-stripe commented:
Hey @wendigo would you please take a look at this PR and provide feedback? thanks!

wendigo commented:
@behrooz-stripe sorry I don't have context here

behrooz-stripe commented:
@electrum might have the best context on this based on discussions in this thread: https://trinodb.slack.com/archives/CP1MUNEUX/p1715720365582019

patstevens4 commented:
Bump, eagerly waiting for this feature! Thanks for all the work on this great idea

behrooz-stripe commented:
Bump @Praveen2112 @hashhar

Praveen2112 commented:
Sorry for the delay. I'll take a look this week.

Praveen2112 commented on core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:
```diff
@@ -1596,7 +1596,33 @@ protected Scope visitQuery(Query node, Optional<Scope> scope)
+            }
+
             analysis.setScope(node, queryScope);
+            ImmutableList<OutputColumn> outputColumns = outputColumnsBuilder.build();
```
nit: List<OutputColumn>

Praveen2112 commented on core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:
```diff
@@ -1596,7 +1596,33 @@ protected Scope visitQuery(Query node, Optional<Scope> scope)
+                        }
+                    }
+                }
+                analysis.setUpdateTarget(version, qualifiedName, Optional.empty(), Optional.of(outputColumns));
```
I'm not sure if using `updateTarget` is a right approach here as it is used for a different set of operations and we can't bind it to the first table we encounter as a part of SELECT query

behrooz-stripe replied:
I did think about this and tried to first add a dedicated target but it required quite a bit of plumbing and changes creeped in across many classes and ultimately the EventListener as well; hence the snippet in the description of the PR. So I wasn't sure if we want to go down a major change. 
I do agree it's somewhat awkward to tie a select to a first table but ultimately binding a select to table is kinda irrelevant conceptually and should be just ignored. I am happy to change course on this as well.

Praveen2112 commented on core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:
```diff
@@ -1596,7 +1596,33 @@ protected Scope visitQuery(Query node, Optional<Scope> scope)
+            if (!outputColumns.isEmpty()) {
+                QualifiedObjectName qualifiedName = new QualifiedObjectName("", "", "");
+                CatalogHandle.CatalogVersion version = new CatalogHandle.CatalogVersion("1");
+                if (node.getQueryBody() instanceof QuerySpecification querySpecification) {
```
This check is a bit brittle as we capture them only for simple queries and not for queries which has `WITH` clause in the begining.

Praveen2112 commented on testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerWithSplits.java:
```diff
@@ -122,7 +123,10 @@ public void testSplitsForNormalQuery()
         assertThat(queryCompletedEvent.getContext().getResourceGroupId().isPresent()).isTrue();
         assertThat(queryCompletedEvent.getContext().getResourceGroupId().get()).isEqualTo(createResourceGroupId("global", "user-user"));
-        assertThat(queryCompletedEvent.getIoMetadata().getOutput()).isEqualTo(Optional.empty());
+        Optional<QueryOutputMetadata> output = queryCompletedEvent.getIoMetadata().getOutput();
```
Instead of capturing this event for `SELECT` in  `QueryOutputMetadata` ? Can we add additional field in `QueryIOMetadata` where there could be a `Optional<List<OutputColumnMetadata>>` which could be set only for `SELECT` statements in this way the logic could be simplified. 

@martint - Is `QueryIOMetadata` a right place to add them or is there a better place we could store them ?

behrooz-stripe replied:
I did take a stab at storing these info separately from `output` and as I [mentioned here](https://github.com/trinodb/trino/pull/23322#discussion_r1904382885) it (changes) quickly leaks to many places (
QueryInfo.java, QueryStateMachine.java, QueryMonitor.java, QueryCompletedEvent.java, ...) to plumb this info from analysis all the way to to event listener.
So basically my idea was to avoid that and reuse update target for carrying the same info. In any case, I put a [gist](https://gist.github.com/behrooz-stripe/5ee72ede9c416e2cfda287003cf41ac2) of the actual diff if I were to hold this in a separate variable [in the analysis](https://gist.github.com/behrooz-stripe/5ee72ede9c416e2cfda287003cf41ac2#file-diffsepfield-txt-L179), and CompletedEvent, and [QueryInfo](https://gist.github.com/behrooz-stripe/5ee72ede9c416e2cfda287003cf41ac2#file-diffsepfield-txt-L69). I am happy to update this PR to reflect that if you folks prefer that. 
CC: @martint @Praveen2112

behrooz-stripe replied:
Hey @Praveen2112 @martint any update here? Thank you!

Praveen2112 replied:
update target is mostly for tables which are modified i.e INSERT or UPDATE or DELETE or MERGE - But in case of SELECT it is a different operation so I think it could be a dedicated field for them.

Praveen2112 commented:
@behrooz-stripe Can you please rebase the PR - I'll review it this week.

behrooz-stripe commented:
Hey @Praveen2112 I was on leave for a while and didn't get a chance to do this earlier. But I have addressed all your comments. Please reopen and let's get this done. Thank you!
Let  me know if you want me to open a new pull request but we'd lose context that way.
